Ensure we detect properly bulk error when creating versions

williamlardier opened the pull request:
## this was tested successfully with the script that reproduces the problem - no more 502/503 nor InternalErrors returned.

We used to rely on a non-working detection algorithm that always
detected that the versionID operation failed, even if it was
the master version. As a result, the retry would recreate the
versionID a second time, and lead to more race conditions and
invalid states.

Now, we properly detect the index of the failing operation.
For `version`: we retry. For `master`: we ignore, it will be
repaired.

Issue: ARSN-491

Seragonia committed:
Ensure we detect properly bulk error when creating versions

We used to rely on a non-working detection algorithm that always
detected that the versionID operation failed, even if it was
the master version. As a result, the retry would recreate the
versionID a second time, and lead to more race conditions and
invalid states.

Now, we properly detect the index of the failing operation.
For versionID: we retry. For mAsters: we ignore, it will be
repaired.

Issue: ARSN-491

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
                         code: err.code,
                         error: err.errmsg,
-                        isRetry: isRetry ? true : false, // eslint-disable-line no-unneeded-ternary
+                        isRetry: isRetry ? true : false,
```
nit: the ternary is only needed (is it really?) for the case where isRetry is undefined [based on the type], so we may do this instead
```suggestion
                        isRetry: isRetry || false,
```
or if we want to be even more resilient:
```suggestion
                        isRetry: !!isRetry,
```

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
-                                this.putObjectVerCase1(c, bucketName, objName, objVal, params, log, cb, true));
+                    
+                    const writeErrors = err.writeErrors || [];
+                    const firstError: WriteError = writeErrors[0];
```
are errors guaranteed to be "ordered" ?
i.e. could we have instead `writeErrors[1].index === 0` and `writeErrors[0].index === 1` ?

williamlardier replied:
No, that's why we rely on the `index` property. the "first" here meant the `[0]` index of the array: if defined, we have at least one error...
I'll rename it better.

francoisferrand replied:
would be worth pointing out (in the comments) that we can only have a single error, as the way we access it (including index) hints that there could be more...

williamlardier replied:
I added a comment.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
+                        }
+                    }
+                    
+                    // If we can't determine which operation failed from writeErrors,
+                    // fall back to checking result counts
+                    const upsertedCount = err.result?.upsertedCount || err.result?.nUpserted || 0;
```
can this really happen, or is it just an happen to cover all possible "code path" ?
is there any information in mongodb regarding the possible outcomes of batch operation?

williamlardier replied:
Logcally it could happen but the documentation (see https://www.mongodb.com/docs/manual/reference/method/db.collection.bulkWrite/#std-label-bulkwrite-error-handling-txn) does not really state it's possible. I looked at the code and `writeErrors` is a `OneOrMore`, meaning if we enter the catch, we are sure to have at least one element... I can remove this logic to simplify.
I don't think a write could silently fail...

williamlardier replied:
Note: I first kept this check "just in case" the isMasterOpError is not true but we really inserted something. But I wonder if that's wise, because then if there was no master version and the operation failed, we end up passing this check and returning a version ID, even if the master was not created, leading to an inconsistent state (if no master, then we cannot list/get the versionID...). So I removed it completely.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
+                    log.error('putObjectVerCase1: race condition upserting versionId', {
+                        error: err.errmsg,
+                    });
+                    return cb(errors.InternalError);
```
overal logic sounds ok, but the code feels too complicated, with nested conditions and duplicated "exits"...
can we rework it in a simpler way, maybe something like:

```
   if (write_error_on_master || (no_write_error && upsertedCount === 1)) {
      // error on master → success
      return cb(null, `{"versionId": "${versionId}"}`);
   }
   if (!isRetry) {
      return nextTick(...);
   }
   return internalError;
```

williamlardier replied:
Logic was simplified.

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
+                    const upsertedCount = err.result?.upsertedCount || err.result?.nUpserted || 0;
+                    
+                    // If we successfully upserted one document (the version), but had an error on master,
+                    // consider this a success
```
I think the original comment gave a clearer context:
```
Otherwise this error is expected, it means that two differents version was put at the same time
```

It is not just that we "consider" it a success, it is really that we expect this to happen (and it's fine) on simultaneous version put : so may be better to somewhat "merge" both comments

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
                 /*
                  * Related to https://jira.mongodb.org/browse/SERVER-14322
                  * It happens when we are pushing two versions "at the same time"
                  * and the master one does not exist. In MongoDB, two threads are
```
> and the master one does not exist

- can this only happen when the master does not exist? (i.e. trying to "create" the object twice at the same time)
- don't we have a similar issue (maybe not in this "case") when updating the master with 2 new revisions pushed at the same time? Can we add tests as for this case as well, if not already covered?

williamlardier replied:
The discussed mongodb defect was already fixed in MongoDB v5.0.
But otherwise yes, any write conflict will create this error. In our case it should not be possible, given that versionIDs are properly implemented.

> when updating the master with 2 new revisions pushed at the same time

If by revision you mean updating the latest version ID concurrently, then yes, it's done in https://github.com/scality/Arsenal/pull/2387/files#diff-743cf32ff59c21d31aec6a69642b8af09c99bed0329c5b5a8772a6bd440e4e6fR3149, where we "ignore" the error here, because it will be repaired

francoisferrand commented on lib/storage/metadata/mongoclient/MongoClientInterface.ts:
```diff
@@ -859,45 +871,71 @@ class MongoClientInterface {
                  * trying to create the same key, the master version, and one of
                  * them, the one with the highest versionID (less recent one),
                  * fails.
-                 * We check here than than the MongoDB error is related to the
-                 * second operation, the master version update and than the error
-                 * code is the one related to mentionned issue.
```
this comment looks somewhat good (except for typos)?
I think it would be good to keep such very high-level overview (i.e. we ignore error on master document, if the version document was successfully written)

williamlardier replied:
I adapted it a bit, because it was really discussing a mongodb defect already fixed. The new version will present the risk of version ID conflict and the two cases: we can fail either at the first, or the second operation, for different reasons.

Seragonia committed:
Simplify bulk error handling logic

- The bulk is guaranteed to return an error, so no need to
  handle two paths, just one is enough, while checking for the
  number of inserted documents.
- Improve documentation.
- Removed tests that are not possible: the bulk error will be
  defined if we enter the .catch.

Issue: ARSN-491

welansari commented on tests/unit/storage/metadata/mongoclient/MongoClientInterface.spec.js:
```diff
@@ -3051,3 +3051,207 @@ describe('MongoClientInterface, deleteBucketIndexes', () => {
+            params,
+            logger,
+            () => {
+                try {
+                    // Check that bulkWrite was called twice (original + retry)
+                    assert.strictEqual(bulkWriteStub.callCount, 2, 'Expected bulkWrite to be called twice');
+
+                    // Verify putObjectVerCase1 was called with isRetry=true for the retry
+                    assert(putObjectVerCase1Spy.calledTwice, 'Expected putObjectVerCase1 to be called twice');
+                    assert(putObjectVerCase1Spy.secondCall.args[7], 'Expected second call to have isRetry=true');
+
+                    done();
+                } catch (assertionError) {
+                    done(assertionError);
+                }
```
nit: do we really need to catch the assertion error ?

francoisferrand commented on tests/unit/storage/metadata/mongoclient/putObject.spec.js:
```diff
@@ -221,28 +221,12 @@ describe('MongoClientInterface:putObjectVerCase1', () => {
     });
 
-    it('should not fail when error code is 11000 and upsertedCount is 1', done => {
+    it('should fail when error code is 11000, master operation failed', done => {
```
- `index === 0` means the "version" document failed, not master 
- should we add a test case here for the "master operation failed" case indeed (which should not fail the operation)

williamlardier replied:
The tests were updated, i also merged the two similar test suites into one, and it covers all cases, in the right file. I think we should either split the MongoClientInterface accordingly, or use a single test file though, as it doesn't follow the usuql pattern to have different names for subparts of one (big) ts file...

francoisferrand commented on tests/unit/storage/metadata/mongoclient/MongoClientInterface.spec.js:
```diff
@@ -3051,3 +3051,195 @@ describe('MongoClientInterface, deleteBucketIndexes', () => {
     });
 });
+
+describe('MongoClientInterface, putObjectVerCase1', () => {
```
should these not go in `tests/unit/storage/metadata/mongoclient/putObject.spec.js`, which already has a `MongoClientInterface:putObjectVerCase1` test suite?

bert-e commented:
# Hello williamlardier,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/arsenal/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Request integration branches



Waiting for integration branch creation to be requested by the user.

To request integration branches, please comment on this pull request with the following command:

```
/create_integration_branches
```

Alternatively, the `/approve` and `/create_pull_requests` commands will automatically
create the integration branches.

Seragonia committed:
Bump Arsenal

Issue: ARSN-491

ghost commented:
/create_integration_branches

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/8.2/bugfix/ARSN-491` with contents from `bugfix/ARSN-491`
and `development/8.2`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/8.2/bugfix/ARSN-491 origin/development/8.2
 $ git merge origin/bugfix/ARSN-491
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/8.2/bugfix/ARSN-491
```





*The following options are set:* **create_integration_branches**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers











*The following options are set:* **create_integration_branches**

ghost commented:
/approve

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.1`

* :heavy_check_mark: `development/8.2`



The following branches have **NOT** changed:

* `development/6.4`
* `development/7.10`
* `development/7.4`
* `development/7.70`



Please check the status of the associated issue ARSN-491.

Goodbye williamlardier.




*The following options are set:* **approve, create_integration_branches**
